XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,300 @@
+                Description = string.Format(Messages.DOWNLOAD_AND_EXTRACT_ACTION_DOWNLOADING_DESC, updateName);
+                LogDescriptionChanges = false;
+                DownloadFile();
+                LogDescriptionChanges = true;
+
+                if (IsCompleted || Cancelled)
+                    return;
+
+                if (Cancelling)
+                    throw new CancelledException();
+            }
+            
+            ValidateMsi();
+
+            // Install the msi            
+            try
+            {
+                if (File.Exists(outputPathAndFileName))
+                {
+                    // Start the install process, it will handle closing of application.
+                    Process.Start(outputPathAndFileName);
+                    log.DebugFormat("Update {0} found and install started", updateName);
+                }
+            }
+            catch (Exception e)
+            {
+                if (File.Exists(outputPathAndFileName))
+                {
+                    File.Delete(outputPathAndFileName);
+                }
+                log.Error("Exception occurred when installing CHC.", e);
+                throw;
+            }
+
+            Description = Messages.COMPLETED;
+            MarkCompleted();
+        }
+
+        private void ValidateMsi()
+        {
+
+            using (FileStream stream = new FileStream(outputPathAndFileName, FileMode.Open, FileAccess.Read))
+            {
+                var hash = StreamUtilities.ComputeHash(stream, out _);
+                
+                //Convert to Hex string
+                StringBuilder sb = new StringBuilder();
+                foreach (var b in hash)
+                    sb.Append(b.ToString("x2"));
+             
+                var calculatedChecksum = sb.ToString();
+                
+                // Check if calculatedChecksum matches what is in chcupdates.xml
+                if (checksum != calculatedChecksum)
+                    throw new Exception("The checksum of the downloaded MSI does not match what is expected. This indicates it has been tampered with.");
+            }
+
+            var valid = false;
+            // Check digital signature of .msi
+            using (var basicSigner = X509Certificate.CreateFromSignedFile(outputPathAndFileName))
+            {
+                using (var cert = new X509Certificate2(basicSigner))
+                {
+                    try
+                    {
+                        valid = cert.Verify();
+                    }
+                    catch (Exception e)
+                    {
+                        throw new Exception("CHC could not validate the certificate associated with the installation file.", e);
+                    }
+                }
+            }
+
+            if (!valid)
+                throw new Exception("Invalid digital signature on msi.");

```

Comment:
This one should also be a message that we save in Messages.resx, since it is shown to the user on failure.